snuba/optimize.py

```diff
@@ -291,15 +359,20 @@ def optimize_partitions(
 
         query = (query_template % args).strip()
         logger.info(f"Optimizing partition: {partition}")
-        start = time.time()
 
         # Update tracker before running clickhouse.execute since its possible
         # that the connection can get disconnected while the server is still
         # executing the optimization. In that case we don't want to
         # run optimization on the same partition twice.
         if tracker:
-            tracker.update_completed_partitions(partition)
+            tracker.update_scheduled_partitions(partition)
+
+        # if theres a merge in progress for this partition, wait for it to finish
+        if partition in current_merge_info:
+            est_sleep_time = current_merge_info[partition]["estimated_time"]
+            time.sleep(OPTIMIZE_BASE_SLEEP_TIME + est_sleep_time)
```

Comment:
I think this algorithm to make the decision is quite weak. 
- It relies on the estimate time to be at least reasonably accurate. I don;'t think we ever measured this accuracy, but if there is an increase in load or a slow down in the disk latency, definitively the estimated time will increase. 
- we run two optimize merges at a time in general. they do not necessarily start at the same time, though one will add load to the system while the other is running thus impacting the estimate.
- The query returns only one of the large merges while we may have mutations going on. 

I think you can make it safer:
- never start anything if there are more than 3 merges on going. 
- ignore all the merges that are expected to take short in total (like 10 minutes or so). 
- If there are long merges on going (>10 minutes), get the expected time, wait till then and repeat the checks.
- make the constants above (3 concurrent merges and min 10 minutes configurable via settings)

So you are sure you are not going to rely on an unreliable estimate and you take into account mutations that may have started in the meantime

Reply:
Agreed with what Filippo has mentioned above. I have commented here https://github.com/getsentry/snuba/pull/3158/files#diff-85a4b5d71565524e6a2d3f9540361a4b0093941afb3ae0a8f94bc0532ff87d49R206 what the api should do. Ideally that api should be used here directly so that it can be used in a loop to keep checking that the count has gone to 0. You don't need to know the exact partition which is having the long merge. You want to know the count of them.

Reply:
Updated the algorithm. We sleep if clickhouse is busy, where Clickhouse is considered busy if:

1. there are more than settings.OPTIMIZE_MERGE_MAX_CONCURRENT_JOBS merges in progress with an elapsed time > settings.OPTIMIZE_MERGE_MIN_ELAPSED_CUTTOFF_TIME. These are merges that we consider long running and should probably wait on before attempting again.
2. Or there is a merge where total_size_bytes_compressed is greater than settings.OPTIMIZE_MERGE_SIZE_CUTOFF. These are merges on large parts and we should wait for them.